designs/0016-hessian_optimize_cmdstan.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: Allow Cmdstan to optionally print out draws from Laplace approximation of posterior when optimizing
+- Start Date: 2020-03-08
+- RFC PR: 16
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+When computing a MAP estimate, the Hessian of the log density can be used to construct a normal approximation to the posterior.
+
+# Motivation
+[motivation]: #motivation
+
+I have a Stan model that is too slow to practically sample all the time. Because optimization seem to give reasonable results, it would be nice to have the normal approximation to the posterior to give some sense of the uncertainty in the problem as well.
+
+An approximate posterior covariance comes from computing the inverse of the Hessian of the negative log density.
+
+Rstan already supports this via the 'hessian' argument to 'optimizing'.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This adds two arguments to the cmdstan interface.
+
+```laplace_draws``` - The number of draws to take from the posterior approximation. By default, this is zero, and no laplace approximation is done
+
+```laplace_diag_shift``` - A value to add to the diagonal of the hessian approximation to fix small non-singularities (defaulting to zero)
+
+The output is printed after the optimimum.
+
+A model can be called by:
+```
+./model optimize laplace_draws=100 data file=data.dat
+```
+
+or with the diagonal shift:
+```
+./model optimize laplace_draws=100 laplace_diag_shift=1e-10 data file=data.dat
+```
+
+Optimizing output currently looks like:
+```
+# stan_version_major = 2
+...
+#   refresh = 100 (Default)
+lp__,b.1,b.2
+3427.64,7.66366,5.33466
+```
+
+The new output would look like:
+
+```
+# stan_version_major = 2
+...
+#   refresh = 100 (Default)
+lp__,b.1,b.2
```

Comment:
can you add `lg__`, ie, log density with respect to the approximation? Then it would be easier to use Pareto k diagnostic and do importance resampling.

Reply:
I can evaluate both the true log density and the log density of the normal approximation. Just want one or both?

Reply:
Both `lp__` and `lg__` as in CmdStan advi, and in RStan advi and Laplace